src/twisted/web/server.py

```diff
@@ -878,3 +884,110 @@ def acceptableProtocols(self):
             baseProtocols.insert(0, b"h2")
 
         return baseProtocols
+
+
+class _HTTPSRedirectProtocolWrapper(ProtocolWrapper):
+    """
+    Wraps the TLS protocol and redirect to HTTPS if an HTTP request is
+    detected after a connection is made.
+    """
+
+    def dataReceived(self, data):
+        """
+        Only called when initial connection data is received over SSL.
+
+        After that, the dataReceived method of the wrapped protocol is called.
+
+        It checks if an HTTP request is made over HTTPS port, instead of an
+        SSL handshake.
+
+        @param data: The data received from the remote HTTP client peer.
+        @type data: L{bytes}.
+        """
+
+        def get_http_host(data):
+            """
+            Return the value of the Host header found inside `data` HTTP
+            raw request.
+
+            @param data: The first data chunk received from the remote HTTP client peer.
+            @type data: L{bytes}.
+
+            @return: The value of the Host header. It returns the empty string when the value could not be extracted.
+            @rtype: L{bytes}
+            """
+            parts = data.split(b"\r\n")
+            if len(parts) == 1:
+                return b""
+
+            for part in parts:
+                if not part.lower().startswith(b"host"):
+                    continue
+
+                segments = part.split(b":", 1)
+                if len(segments) != 2:
+                    return b""
+
+                return segments[1].strip()
+
+            return b""
+
+        # We only want to handle the first package.
+        self.dataReceived = self.wrappedProtocol.dataReceived
+
+        host = get_http_host(data)
+
+        if not host:
+            # Just forward the data to the wrapped protocol.
+            return self.dataReceived(data)
+
+        self.transport.write(
```

Comment:
We should *really* emit a noisy log message here, since this is a bad and very odd misconfiguration that you should probably try to fix.  Maybe log something about the client's configuration?

Reply:
Well, the client configuration in this use case is the "human brain" 

The user types 'service.example.com:1234` in the browser  and the browser makes an HTTP request.
The fix for  the user to remember to type `https://service.example.com:1234`

I don't know how to fix the human brain so this is why I have not emitted a log.
But I think that a log should be emitted , to detect misconfigurations made by automated requests.